File: C/statistics.c
#include <stdio.h>
#include <stdlib.h>

int compare_ints(const void *a, const void *b) {
    int ia = *(const int *)a;
    int ib = *(const int *)b;
    return (ia > ib) - (ia < ib);
}

double calculate_mean(const int arr[], int n) {
    long sum = 0;
    for (int i = 0; i < n; i++) {
        sum += arr[i];
    }
    return (double)sum / n;
}

double calculate_median(const int arr[], int n) {
    int *copy = (int *)malloc(n * sizeof(int));
    if (copy == NULL) {
        fprintf(stderr, "Memory allocation failed.\n");
        exit(EXIT_FAILURE);
    }

    for (int i = 0; i < n; i++) {
        copy[i] = arr[i];
    }

    qsort(copy, n, sizeof(int), compare_ints);

    double median;
    if (n % 2 == 1) {
        median = copy[n / 2];
    } else {
        median = (copy[(n / 2) - 1] + copy[n / 2]) / 2.0;
    }

    free(copy);
    return median;
}

void calculate_mode(const int arr[], int n) {
    int *copy = (int *)malloc(n * sizeof(int));
    if (copy == NULL) {
        fprintf(stderr, "Memory allocation failed.\n");
        exit(EXIT_FAILURE);
    }

    for (int i = 0; i < n; i++) {
        copy[i] = arr[i];
    }

    qsort(copy, n, sizeof(int), compare_ints);

    int max_count = 1;
    int current_count = 1;

    for (int i = 1; i < n; i++) {
        if (copy[i] == copy[i - 1]) {
            current_count++;
        } else {
            if (current_count > max_count) {
                max_count = current_count;
            }
            current_count = 1;
        }
    }
    if (current_count > max_count) {
        max_count = current_count;
    }

    if (max_count == 1) {
        printf("Mode: No mode (all values occur once)\n");
        free(copy);
        return;
    }

    printf("Mode(s): ");
    current_count = 1;
    int first_printed = 1;

    for (int i = 1; i <= n; i++) {
        if (i < n && copy[i] == copy[i - 1]) {
            current_count++;
        } else {
            if (current_count == max_count) {
                if (!first_printed) {
                    printf(", ");
                }
                printf("%d", copy[i - 1]);
                first_printed = 0;
            }
            current_count = 1;
        }
    }

    printf(" (frequency = %d)\n", max_count);
    free(copy);
}

void print_list(const int arr[], int n) {
    printf("[");
    for (int i = 0; i < n; i++) {
        printf("%d", arr[i]);
        if (i < n - 1) {
            printf(", ");
        }
    }
    printf("]\n");
}

int main(void) {
    int values[] = {4, 1, 2, 2, 3, 4, 5, 4, 2, 6, 7, 2, 4};
    int n = sizeof(values) / sizeof(values[0]);

    printf("C (Procedural) Statistics Calculator\n");
    printf("Input values: ");
    print_list(values, n);
    printf("Mean: %.2f\n", calculate_mean(values, n));
    printf("Median: %.2f\n", calculate_median(values, n));
    calculate_mode(values, n);

    return 0;
}

File: ocaml/statistics.ml
(* OCaml (Functional) Statistics Calculator *)

let values = [4; 1; 2; 2; 3; 4; 5; 4; 2; 6; 7; 2; 4]

let mean lst =
  let sum = List.fold_left ( + ) 0 lst in
  float_of_int sum /. float_of_int (List.length lst)

let sorted lst = List.sort compare lst

let median lst =
  let s = sorted lst in
  let n = List.length s in
  if n mod 2 = 1 then
    float_of_int (List.nth s (n / 2))
  else
    let a = List.nth s ((n / 2) - 1) in
    let b = List.nth s (n / 2) in
    float_of_int (a + b) /. 2.0

let frequencies lst =
  List.fold_left
    (fun acc x ->
      let count = try List.assoc x acc with Not_found -> 0 in
      (x, count + 1) :: List.remove_assoc x acc)
    [] lst

let mode lst =
  let freq = frequencies lst in
  let max_count = List.fold_left (fun m (_, c) -> max m c) 0 freq in
  if max_count <= 1 then
    ([], max_count)
  else
    let modes =
      freq
      |> List.filter (fun (_, c) -> c = max_count)
      |> List.map fst
      |> List.sort compare
    in
    (modes, max_count)

let string_of_int_list lst =
  "[" ^ String.concat "; " (List.map string_of_int lst) ^ "]"

let string_of_modes modes =
  String.concat ", " (List.map string_of_int modes)

let () =
  Printf.printf "OCaml (Functional) Statistics Calculator\n";
  Printf.printf "Input values: %s\n" (string_of_int_list values);
  Printf.printf "Mean: %.2f\n" (mean values);
  Printf.printf "Median: %.2f\n" (median values);
  let modes, freq = mode values in
  match modes with
  | [] -> Printf.printf "Mode: No mode (all values occur once)\n"
  | _ -> Printf.printf "Mode(s): %s (frequency = %d)\n" (string_of_modes modes) freq
